### ISSUE TITLE
Undefined behavior in murmurhashneutral2.c

### ISSUE THREAD

**Issue by noloader** [ccache/ccache#151]
We are testing [Zcash](https://github.com/zcash/zcash). Zcash uses ccache. We added `-fsanitize=undefined` to `CFLAGS` and `CXXFLAGS` and then ran the self tests to see if there were issues. Below is a sample of what we experienced. It appears `murmurhashneutral2.c:18:16` is a problem.

Here's the problematic code.

```
const unsigned char *data = (const unsigned char *)key;
unsigned int k = data[0];
    k |= data[1] << 8;
    k |= data[2] << 16;
    k |= data[3] << 24;
```

The code above is subject to removal by the compiler. At this point, its not clear to me if analysis will identify it and remove it. Intel's ICC is particularly ruthless about removing undefined behavior.

Maybe it would be easier to use one of the swap functions to perform the endian-reversal (if that is what is happening). Or maybe cast the intermediate `data[n]` to an `unsigned int`.

-----

```
/home/zcash/depends/x86_64-unknown-linux-gnu/share/../native/bin/ccache g++ -m64 -DHAVE_CONFIG_H -I. -I../src/config  -I. -I./obj -pthread -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include -I./leveldb/include -I./leveldb/helpers/memenv -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I./secp256k1/include  -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/libsnark -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -std=c++11 -pipe -O1 -fsanitize=undefined -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -MT zcashd-bitcoind.o -MD -MP -MF .deps/zcashd-bitcoind.Tpo -c -o zcashd-bitcoind.o `test -f 'bitcoind.cpp' || echo './'`bitcoind.cpp
murmurhashneutral2.c:18:16: runtime error: left shift of 140 by 24 places cannot be represented in type 'int'
mv -f .deps/zcashd-bitcoind.Tpo .deps/zcashd-bitcoind.Po
/home/zcash/depends/x86_64-unknown-linux-gnu/share/../native/bin/ccache g++ -m64 -DHAVE_CONFIG_H -I. -I../src/config  -I. -I./obj -pthread -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include -I./leveldb/include -I./leveldb/helpers/memenv -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I./secp256k1/include   -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/libsnark -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -std=c++11 -pipe -O1 -fsanitize=undefined -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -MT libbitcoin_server_a-sendalert.o -MD -MP -MF .deps/libbitcoin_server_a-sendalert.Tpo -c -o libbitcoin_server_a-sendalert.o `test -f 'sendalert.cpp' || echo './'`sendalert.cpp
murmurhashneutral2.c:18:16: runtime error: left shift of 155 by 24 places cannot be represented in type 'int'
mv -f .deps/libbitcoin_server_a-sendalert.Tpo .deps/libbitcoin_server_a-sendalert.Po
/home/zcash/depends/x86_64-unknown-linux-gnu/share/../native/bin/ccache g++ -m64 -DHAVE_CONFIG_H -I. -I../src/config  -I. -I./obj -pthread -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include -I./leveldb/include -I./leveldb/helpers/memenv -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I./secp256k1/include   -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/libsnark -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -std=c++11 -pipe -O1 -fsanitize=undefined -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -MT libbitcoin_server_a-addrman.o -MD -MP -MF .deps/libbitcoin_server_a-addrman.Tpo -c -o libbitcoin_server_a-addrman.o `test -f 'addrman.cpp' || echo './'`addrman.cpp
murmurhashneutral2.c:18:16: runtime error: left shift of 155 by 24 places cannot be represented in type 'int'
mv -f .deps/libbitcoin_server_a-addrman.Tpo .deps/libbitcoin_server_a-addrman.Po
/home/zcash/depends/x86_64-unknown-linux-gnu/share/../native/bin/ccache g++ -m64 -DHAVE_CONFIG_H -I. -I../src/config  -I. -I./obj -pthread -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include -I./leveldb/include -I./leveldb/helpers/memenv -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I./secp256k1/include   -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/libsnark -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -std=c++11 -pipe -O1 -fsanitize=undefined -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -MT libbitcoin_server_a-alert.o -MD -MP -MF .deps/libbitcoin_server_a-alert.Tpo -c -o libbitcoin_server_a-alert.o `test -f 'alert.cpp' || echo './'`alert.cpp
murmurhashneutral2.c:18:16: runtime error: left shift of 155 by 24 places cannot be represented in type 'int'
mv -f .deps/libbitcoin_server_a-alert.Tpo .deps/libbitcoin_server_a-alert.Po
/home/zcash/depends/x86_64-unknown-linux-gnu/share/../native/bin/ccache g++ -m64 -DHAVE_CONFIG_H -I. -I../src/config  -I. -I./obj -pthread -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include -I./leveldb/include -I./leveldb/helpers/memenv -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I/home/zcash/depends/x86_64-unknown-linux-gnu/include -I./secp256k1/include   -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/libsnark -I/home/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -std=c++11 -pipe -O1 -fsanitize=undefined -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -MT libbitcoin_server_a-asyncrpcoperation.o -MD -MP -MF .deps/libbitcoin_server_a-asyncrpcoperation.Tpo -c -o libbitcoin_server_a-asyncrpcoperation.o `test -f 'asyncrpcoperation.cpp' || echo './'`asyncrpcoperation.cpp
murmurhashneutral2.c:18:16: runtime error: left shift of 155 by 24 places cannot be represented in type 'int'
...
```

**Comment by afbjorklund** [comment]
Think you are right, there are 3 places where "undefined" would show warnings.
This particular file/problem should probably be reported back upstream, as well.

**Comment by afbjorklund** [comment]
Fixed in 44ee7dae59adb3ba46742dd8328efcaa30b9f25e (added explicit cast).

**Comment by jrosdahl** [comment]
Should be fixed on 3.3-maint and master now by @afbjorklund's fixes.
